### PR TITLE
codeintel: Fix LSIF -> SCIP range conversion

### PR DESCRIPTION
**Pull request by efritz**
This PR makes the following changes. Please review by commit.

- Replace `DefinitionMatcher` callback with a more general `TargetRangeFetcher`. The former returns true if one of the targets of a result set matches the given range ID. The later just returns the set of target rages. This can be used for the same purpose but also helps us with checking for ranges related by implementations relationships in LSIF.
- Read from result chunks containing either a definition result or an implementations result, not only def results.
- Deduplicate symbol information when converting LSIF -> SCIP
- Rewrite `convertRange` that converts an LSIF range into a set of SCIP occurrences and symbol information metadata. The new version now looks _through_ result sets to the target range ID and uses that. We no longer directly encode result set identifiers into migrated LSIF indexes.

## Test plan

Tested implementations by hand in larger branch.

### CODE REVIEW

File: lib/codeintel/lsif/scip/document.go [hunk 1]
```diff
@@ -56,14 +54,15 @@ func ConvertLSIFDocument(
...
 
-			for _, other := range symbol.implementationRelationships {
-				if _, ok := implementationsBySymbolName[symbol.name]; !ok {
-					implementationsBySymbolName[symbol.name] = map[string]struct{}{}
+			for _, interfaceName := range symbol.implementationRelationships {
+				if _, ok := interfacesBySymbolName[symbol.name]; !ok {
+					interfacesBySymbolName[symbol.name] = map[string]struct{}{}
 				}
 
-				implementationsBySymbolName[symbol.name][other] = struct{}{}
```
Comment (efritz) [inline lib/codeintel/lsif/scip/document.go:66]: was never used; new block below adds relationships to symbols

File: lib/codeintel/lsif/scip/document.go [hunk 2]
```diff
@@ -110,64 +135,13 @@ type symbolMetadata struct {
...
-			Symbol:      symbolName,
-			SymbolRoles: int32(symbolRoles),
-		})
-
-		symbols = append(symbols, symbolMetadata{
-			name:                        symbolName,
-			documentation:               document.HoverResults[r.HoverResultID],
-			implementationRelationships: implementsSymbolNames,
-		})
-	}
```
Comment (efritz) [inline lib/codeintel/lsif/scip/document.go:149]: This was lifted into the inline `addOccurrence` defined below

File: lib/codeintel/lsif/scip/document.go [hunk 3]
```diff
@@ -199,15 +173,74 @@ func constructSymbolNames(uploadID int, document precise.DocumentData, r precise
...
+
+		// Add definition of each moniker
+		for _, moniker := range monikers {
+			addOccurrence(moniker, role)
+		}
+
+		// Add definition of each implements relationship
+		for _, moniker := range implementsMonikers {
+			addOccurrence(moniker, role)
+		}
```
Comment (efritz) [inline lib/codeintel/lsif/scip/document.go:228]: If we're a definition range, add `{uploadID}:{rangeID}` `{moniker}` and `{implements}` definition symbols.

File: lib/codeintel/lsif/scip/document.go [hunk 4]
```diff
@@ -199,15 +173,74 @@ func constructSymbolNames(uploadID int, document precise.DocumentData, r precise
...
+
+		for _, targetRangeID := range targetRangeFetcher(r.DefinitionResultID) {
+			// Add reference to the defining range identifier
+			addOccurrence(constructSymbolName(uploadID, targetRangeID), role)
+
+			// Add reference to each moniker
+			for _, moniker := range monikers {
+				addOccurrence(moniker, role)
+			}
+		}
```
Comment (efritz) [inline lib/codeintel/lsif/scip/document.go:240]: If we're not a definition, add `{uploadID}:{definitionRangeID}` and `{moniker}` reference symbols.

Comment (efritz) [reply lib/codeintel/lsif/scip/document.go:240]: Note: each symbol shares the same documentation and implements relationship (see `addOccurrence` above).

File: lib/codeintel/lsif/scip/document.go [hunk 5]
```diff
@@ -199,15 +173,74 @@ func constructSymbolNames(uploadID int, document precise.DocumentData, r precise
...
+			monikers = append(monikers, symbolName)
 		case "implementation":
-			implementsSymbolNameMap[symbolName] = struct{}{}
+			implementsMonikers = append(implementsMonikers, symbolName)
+		}
+	}
+
+	for _, targetRangeID := range targetRangeFetcher(r.ImplementationResultID) {
+		implementsMonikers = append(implementsMonikers, constructSymbolName(uploadID, targetRangeID))
+	}
```
Comment (efritz) [inline lib/codeintel/lsif/scip/document.go:186]: Construct symbol names like `{uploadID}:{targetRangeID}`.

**Review by Numbers88s — APPROVED** [review]

**Pull request merged**